migrations/1686291754032-copy-legacy-tools-to-ctl.js

```diff
@@ -0,0 +1,374 @@
+			config_secret: ltiToolTemplate.secret,
+			config_ressource_link_id: ltiToolTemplate.ressource_link_id,
+			config_lti_message_type: ltiToolTemplate.lti_message_type,
+			config_privacy_permission: ltiToolTemplate.privacy_permission,
+		};
+	}
+
+	return toolConfig;
+}
+
+function mapToExternalTool(ltiToolTemplate) {
+	return {
+		name: ltiToolTemplate.name,
+		url: ltiToolTemplate.url,
+		logoUrl: ltiToolTemplate.logo_url,
+		parameters: mapToExternalToolParameter(ltiToolTemplate),
+		isHidden: ltiToolTemplate.isHidden,
+		openNewTab: ltiToolTemplate.openNewTab,
+		version: 1,
+		...toolConfigMapper(ltiToolTemplate),
+	};
+}
+
+function mapToSchoolExternalTool(externalTool, ltiToolTemplate, context) {
+	return {
+		tool: externalTool._id,
+		school: context.schoolId,
+		schoolParameters: mapToCustomParameterEntry(externalTool.parameters, ltiToolTemplate.customes),
+		toolVersion: externalTool.version,
+	};
+}
+
+function mapToContextExternalTool(schoolExternalTool, externalTool, context) {
+	return {
+		schoolTool: schoolExternalTool._id,
+		contextId: context.id,
+		contextType: 'course',
+		contextToolName: externalTool.name,
+		parameters: schoolExternalTool.parameters,
+		toolVersion: schoolExternalTool.version,
+	};
+}
+
+// How to use more than one schema per collection on mongodb
+// https://stackoverflow.com/questions/14453864/use-more-than-one-schema-per-collection-on-mongodb
+module.exports = {
+	up: async function up() {
+		await connect();
+
+		// eslint-disable-next-line no-process-env
+		if (process.env.SC_THEME !== 'n21') {
+			info('Migration does not migrate lti tools to CTL for this instance.');
+			return;
+		}
+
+		// FIND ALL LTI TOOL TEMPLATES
+		const ltiToolTemplates = await LtiTool.find({
+			$or: [{ name: { $regex: `Bettermarks` } }, { name: { $regex: `Nextcloud` } }],
+			isTemplate: true,
+		})
+			.lean()
+			.exec();
+
+		if ((ltiToolTemplates || []).length === 0) {
+			alert('No LtiTool Template found.');
+			return;
+		}
+
+		// FIND ALL LEGACY TOOLS
+		const ltiTools = await LtiTool.find({
+			$or: [{ name: { $regex: `Bettermarks` } }, { name: { $regex: `Nextcloud` } }],
+			isTemplate: false,
+		})
+			.lean()
+			.exec();
+
+		/* eslint-disable no-await-in-loop */
+		for (const ltiTool of ltiTools) {
+			// GET TOOLTEMPLATE
+			const toolTemplate = ltiToolTemplates.filter((template) => template.name === ltiTool.name);
+
+			// GET COURSE
+			const course = await Course.findOne({
+				ltiToolIds: { $in: [ltiTool.id] },
+			})
+				.lean()
+				.exec();
+
+			if (course === undefined) {
+				alert(`No course found with LtiToolId: ${ltiTool.id}.`);
+				// eslint-disable-next-line no-continue
+				continue;
+			}
+
+			// GET EXTERNALTOOL
+			let externalTool = await ExternalTool.findOne({
+				name: toolTemplate.name,
+				url: toolTemplate.url, // or other paramter for the query
+			})
+				.lean()
+				.exec();
+
+			// CHECK IF EXTERNALTOOL EXISTS
+			if (externalTool === undefined) {
```

Comment:
I think having a structure where you start to migrate all lti templates to external tools first and then iterate over the courses and find missing ones would be better, since you want the info from the template to be the primary source for the external tool.